File: desktop/LegionControl.Desktop/Config/ControllerEditor.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using LegionControl.Desktop.Contract;

namespace LegionControl.Desktop.Config;

/// A change to the setup document, worked out in full before anything is written.
///
/// Every edit here produces one of these first: the document as it would be, what this app would
/// refuse about it, and what it would change in the machines the user can see. Nothing is written
/// from a control that did not first show this.
public sealed record EditPlan(
    string Description,
    string Text,
    IReadOnlyList<string> Problems,
    SetupConflictPreview Preview)
{
    public bool IsApplicable => Problems.Count == 0;
}

/// Structured edits to the setup document, made on the document as it is rather than on the model
/// this app decoded from it.
///
/// The distinction matters: the document may carry keys written for the phone, for a future agent,
/// or by hand, and re-serialising this app's own idea of the setup would quietly drop every one of
/// them. So the JSON tree is edited in place and everything untouched stays exactly as it was,
/// including the user's key order.
public static class ControllerEditor
{
    private static readonly JsonSerializerOptions Writing = new() { WriteIndented = true };

    /// Applies a change to the document text and works out what it would mean.
    public static EditPlan Plan(string currentText, ControllerConfig? current, string description, Action<JsonObject> change)
    {
        JsonObject root;
        try
        {
            root = JsonNode.Parse(string.IsNullOrWhiteSpace(currentText) ? "{}" : currentText) as JsonObject
                   ?? new JsonObject();
        }
        catch (JsonException error)
        {
            return new EditPlan(description, currentText, new[] { $"The document is not valid JSON: {error.Message}" },
                SetupConflictPreview.Between(current, new ControllerConfig()));
        }

        try
        {
            change(root);
        }
        catch (Exception error)
        {
            return new EditPlan(description, currentText, new[] { error.Message },
                SetupConflictPreview.Between(current, new ControllerConfig()));
        }

        var text = root.ToJsonString(Writing);
        var decoded = ControllerConfig.From(Value.Parse(text));
        return new EditPlan(description, text, decoded.Problems(), SetupConflictPreview.Between(current, decoded));
    }

    /// Adds or replaces one machine.
    public static Action<JsonObject> UpsertMachine(JsonObject machine) => root =>
    {
        var id = machine["id"]?.GetValue<string>();
        if (string.IsNullOrWhiteSpace(id)) throw new InvalidOperationException("A machine needs an id.");
        var machines = Array(root, "machines");
        for (var index = 0; index < machines.Count; index += 1)
        {
            if (machines[index] is JsonObject existing && existing["id"]?.GetValue<string>() == id)
            {
                machines[index] = machine.DeepClone();
                return;
            }
        }
        machines.Add(machine.DeepClone());
    };

    public static Action<JsonObject> RemoveMachine(string id) => root =>
    {
        var machines = Array(root, "machines");
        for (var index = machines.Count - 1; index >= 0; index -= 1)
        {
            if (machines[index] is JsonObject existing && existing["id"]?.GetValue<string>() == id)
            {
                machines.RemoveAt(index);
            }
        }
    };

    public static Action<JsonObject> UpsertSystem(string machineId, JsonObject system) => root =>
    {
        var machine = FindMachine(root, machineId);
        var id = system["id"]?.GetValue<string>();
        if (string.IsNullOrWhiteSpace(id)) throw new InvalidOperationException("A system needs an id.");
        var systems = Array(machine, "systems");
        for (var index = 0; index < systems.Count; index += 1)
        {
            if (systems[index] is JsonObject existing && existing["id"]?.GetValue<string>() == id)
            {
                systems[index] = system.DeepClone();
                return;
            }
        }
        systems.Add(system.DeepClone());
    };

    public static Action<JsonObject> RemoveSystem(string machineId, string systemId) => root =>
    {
        var machine = FindMachine(root, machineId);
        var systems = Array(machine, "systems");
        for (var index = systems.Count - 1; index >= 0; index -= 1)
        {
            if (systems[index] is JsonObject existing && existing["id"]?.GetValue<string>() == systemId)
            {
                systems.RemoveAt(index);
            }
        }
    };

    public static Action<JsonObject> UpsertEndpoint(string machineId, JsonObject endpoint) => root =>
    {
        var machine = FindMachine(root, machineId);
        var host = endpoint["host"]?.GetValue<string>();
        if (string.IsNullOrWhiteSpace(host)) throw new InvalidOperationException("An address needs a host.");
        var id = endpoint["id"]?.GetValue<string>() ?? host;
        var endpoints = Array(machine, "endpoints");
        for (var index = 0; index < endpoints.Count; index += 1)
        {
            if (endpoints[index] is JsonObject existing && (existing["id"]?.GetValue<string>() ?? existing["host"]?.GetValue<string>()) == id)
            {
                endpoints[index] = endpoint.DeepClone();
                return;
            }
        }
        endpoints.Add(endpoint.DeepClone());
    };

    public static Action<JsonObject> RemoveEndpoint(string machineId, string endpointId) => root =>
    {
        var machine = FindMachine(root, machineId);
        var endpoints = Array(machine, "endpoints");
        for (var index = endpoints.Count - 1; index >= 0; index -= 1)
        {
            if (endpoints[index] is JsonObject existing
                && (existing["id"]?.GetValue<string>() ?? existing["host"]?.GetValue<string>()) == endpointId)
            {
                endpoints.RemoveAt(index);
            }
        }
    };

    public static Action<JsonObject> UpsertSite(JsonObject site) => root =>
    {
        var id = site["id"]?.GetValue<string>();
        if (string.IsNullOrWhiteSpace(id)) throw new InvalidOperationException("A site needs an id.");
        var sites = Array(root, "sites");
        for (var index = 0; index < sites.Count; index += 1)
        {
            if (sites[index] is JsonObject existing && existing["id"]?.GetValue<string>() == id)
            {
                sites[index] = site.DeepClone();
                return;
            }
        }
        sites.Add(site.DeepClone());
    };

    public static Action<JsonObject> RemoveSite(string id) => root =>
    {
        var sites = Array(root, "sites");
        for (var index = sites.Count - 1; index >= 0; index -= 1)
        {
            if (sites[index] is JsonObject existing && existing["id"]?.GetValue<string>() == id) sites.RemoveAt(index);
        }
    };

    public static Action<JsonObject> SetMachineSite(string machineId, string? siteId) => root =>
    {
        var machine = FindMachine(root, machineId);
        if (siteId is null) machine.Remove("site");
        else machine["site"] = siteId;
    };

    public static Action<JsonObject> SetAlwaysOn(string machineId, bool? alwaysOn) => root =>
    {
        var machine = FindMachine(root, machineId);
        if (alwaysOn is null) machine.Remove("alwaysOn");
        else machine["alwaysOn"] = alwaysOn.Value;
    };

    /// Writes the ordered helper list, and the singular alias beside it.
    ///
    /// Both, always. A client from before the list exists reads only `helper`, and leaving it
    /// behind would silently take the wake away from every device that has not been updated yet.
    public static Action<JsonObject> SetWakeHelpers(string machineId, IReadOnlyList<WakeHelper> helpers) => root =>
    {
        var machine = FindMachine(root, machineId);
        if (machine["wake"] is not JsonObject wake)
        {
            throw new InvalidOperationException($"Machine \"{machineId}\" has no wake block to put helpers in.");
        }
        if (helpers.Count == 0)
        {
            wake.Remove("helpers");
            wake.Remove("helper");
            return;
        }
        wake["helpers"] = new JsonArray(helpers
            .Select(helper => (JsonNode)new JsonObject
            {
                ["machine"] = helper.Machine,
                ["action"] = helper.Action,
            })
            .ToArray());
        wake["helper"] = new JsonObject
        {
            ["machine"] = helpers[0].Machine,
            ["action"] = helpers[0].Action,
        };
    };

    /// Takes this device's private settings out of the shared document.
    ///
    /// Only ever run when a person asks for it. Removing a key changes the bytes, and therefore the
    /// hash, so doing it automatically would make one device's copy differ from everyone else's for
    /// a reason nobody chose.
    public static Action<JsonObject> MovePrivateSettingsOut() => root =>
    {
        root.Remove("local");
        foreach (var entry in Array(root, "machines"))
        {
            if (entry is not JsonObject machine) continue;
            if (machine["ssh"] is JsonObject ssh)
            {
                ssh.Remove("identityFile");
                if (ssh.Count == 0) machine.Remove("ssh");
            }
        }
    };

    private static JsonObject FindMachine(JsonObject root, string id)
    {
        foreach (var entry in Array(root, "machines"))
        {
            if (entry is JsonObject machine && machine["id"]?.GetValue<string>() == id) return machine;
        }
        throw new InvalidOperationException($"There is no machine \"{id}\" in the document.");
    }

    private static JsonArray Array(JsonObject owner, string name)
    {
        if (owner[name] is JsonArray existing) return existing;
        var created = new JsonArray();
        owner[name] = created;
        return created;
    }
}

File: mac/Sources/LegionControlCore/ControllerEditor.swift
import Foundation

/// Editing the shared setup from inside the app.
///
/// The file stays the source of truth and stays hand editable: this writes JSON a person can read
/// and never a serialised form of the app's own model. What it adds is the bookkeeping a hand edit
/// cannot do — a revision that counts up and, more importantly, the lineage that lets every other
/// device tell "this descends from what I hold" from "this is a different branch".
///
/// The document is edited as plain JSON rather than through `ControllerConfig`. A setup may
/// legitimately carry keys this build has never heard of: a newer client wrote them, or only the
/// phone reads them. Round-tripping through the typed model would delete every one of them, silently
/// and on one device only.
enum ControllerEditor {

    struct EditFailure: Error, Sendable, Equatable {
        var message: String
        init(_ message: String) { self.message = message }
    }

    /// The result of an edit: what to write, and what the identity became.
    struct Edited: Sendable, Equatable {
        var bytes: Data
        var identity: ControllerIdentity
    }

    /// Apply a change and stamp the new identity on it.
    ///
    /// `parentHash` is the canonical hash of the document being edited, and it becomes the first
    /// entry of the new lineage. That is the whole mechanism: a machine holding the parent sees its
    /// own hash in the lineage of what arrives and accepts it as a fast-forward, and a machine
    /// holding something else does not, and says so instead of being overwritten.
    static func apply(
        to raw: Data,
        parentHash: String,
        parentIdentity: ControllerIdentity,
        deviceName: String,
        now: Date = Date(),
        newIdentifier: @Sendable () -> String = { "setup-" + UUID().uuidString.lowercased() },
        change: (inout [String: Any]) throws -> Void
    ) throws -> Edited {
        guard var root = (try? JSONSerialization.jsonObject(with: raw)) as? [String: Any] else {
            throw EditFailure("The config file is not a JSON object, so it cannot be edited from here. Fix it in an editor first.")
        }

        try change(&root)

        // The identity is stamped after the change rather than before, so a change that threw leaves
        // the document and its lineage exactly as they were.
        let existing = (root["controller"] as? [String: Any]) ?? [:]
        var identity = parentIdentity

        // A document that never had an identity gets one now, and only now: an id invented on every
        // load would make two devices reading the same file believe they were on different setups.
        if identity.id == nil {
            identity.id = (existing["id"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? newIdentifier()
        }
        if identity.name == nil { identity.name = existing["name"] as? String }

        let next = identity.next(parent: parentHash, device: deviceName, now: now)
        root["controller"] = SetupMerge.encode(next)

        let bytes: Data
        do {
            bytes = try JSONText.canonicalDocument(root)
        } catch {
            throw EditFailure("The edited config could not be written back as JSON. \(error.localizedDescription)")
        }

        return Edited(bytes: bytes, identity: next)
    }

    /// Give a document that has never had one an identity, without otherwise touching it.
    ///
    /// This is what happens to a hand written file or a pasted example the first time the app reads
    /// it: it becomes a setup of its own, with a fresh id, revision 1 and no ancestors. It is not a
    /// claim about anybody else's setup, and if the machines turn out to hold another one the user
    /// gets the ordinary "different setup" question.
    static func adopting(
        _ raw: Data,
        deviceName: String,
        setupName: String? = nil,
        now: Date = Date(),
        newIdentifier: @Sendable () -> String = { "setup-" + UUID().uuidString.lowercased() }
    ) throws -> Edited {
        guard var root = (try? JSONSerialization.jsonObject(with: raw)) as? [String: Any] else {
            throw EditFailure("The config file is not a JSON object, so it cannot be given an identity.")
        }
        let identity = ControllerIdentity(
            id: newIdentifier(),
            name: setupName ?? (root["controller"] as? [String: Any])?["name"] as? String,
            revision: 1,
            updatedAt: ISO8601DateFormatter.lenient.string(from: now),
            source: ControllerIdentity.sourceKind,
            device: deviceName,
            lineage: []
        )
        root["controller"] = SetupMerge.encode(identity)
        return Edited(bytes: try JSONText.canonicalDocument(root), identity: identity)
    }

    // MARK: - The changes the app knows how to make

    /// Add or replace one machine, keeping every key of an existing entry the app does not manage.
    static func upsertMachine(_ machine: [String: Any], in root: inout [String: Any]) throws {
        guard let id = machine["id"] as? String, !id.isEmpty else {
            throw EditFailure("A machine needs an id.")
        }
        guard AgentToken.isValid(id) else {
            throw EditFailure("\"\(id)\" is not a usable machine id. Use letters, digits and . _ - : @ / \\ ~ = +.")
        }
        var machines = (root["machines"] as? [[String: Any]]) ?? []
        if let index = machines.firstIndex(where: { $0["id"] as? String == id }) {
            var merged = machines[index]
            for (key, value) in machine {
                if value is NSNull { merged.removeValue(forKey: key) } else { merged[key] = value }
            }
            machines[index] = merged
        } else {
            machines.append(machine)
        }
        root["machines"] = machines
    }

    static func removeMachine(id: String, in root: inout [String: Any]) throws {
        let machines = (root["machines"] as? [[String: Any]]) ?? []
        let remaining = machines.filter { $0["id"] as? String != id }
        guard remaining.count != machines.count else {
            throw EditFailure("There is no machine called \"\(id)\" in the file.")
        }
        root["machines"] = remaining
    }

    /// Add or replace one system on one machine.
    static func upsertSystem(_ system: [String: Any], onMachine machineId: String, in root: inout [String: Any]) throws {
        try mutateMachine(machineId, in: &root) { machine in
            try upsertKeyed(system, in: &machine, list: "systems", what: "system")
        }
    }

    static func removeSystem(id: String, fromMachine machineId: String, in root: inout [String: Any]) throws {
        try mutateMachine(machineId, in: &root) { machine in
            var systems = (machine["systems"] as? [[String: Any]]) ?? []
            systems.removeAll { $0["id"] as? String == id }
            guard !systems.isEmpty else {
                throw EditFailure("Every machine needs at least one system, so the last one cannot be removed.")
            }
            machine["systems"] = systems
        }
    }

    /// Add or replace one endpoint on one machine.
    static func upsertEndpoint(_ endpoint: [String: Any], onMachine machineId: String, in root: inout [String: Any]) throws {
        try mutateMachine(machineId, in: &root) { machine in
            try upsertKeyed(endpoint, in: &machine, list: "endpoints", what: "endpoint")
        }
    }

    static func removeEndpoint(id: String, fromMachine machineId: String, in root: inout [String: Any]) throws {
        try mutateMachine(machineId, in: &root) { machine in
            var endpoints = (machine["endpoints"] as? [[String: Any]]) ?? []
            endpoints.removeAll { $0["id"] as? String == id }
            machine["endpoints"] = endpoints
        }
    }

    /// Which site a machine is at, or nil to unplace it.
    static func setSite(_ site: String?, onMachine machineId: String, in root: inout [String: Any]) throws {
        try mutateMachine(machineId, in: &root) { machine in
            if let site, !site.isEmpty { machine["site"] = site } else { machine.removeValue(forKey: "site") }
        }
    }

    static func setAlwaysOn(_ alwaysOn: Bool, onMachine machineId: String, in root: inout [String: Any]) throws {
        try mutateMachine(machineId, in: &root) { machine in
            machine["alwaysOn"] = alwaysOn
        }
    }

    /// Replace the ordered helper list for one machine.
    ///
    /// Both keys are written: `helpers` for anything that understands failover, and `helper` set to
    /// the first of them so a 1.2 client keeps waking the machine the one way it knows. Clearing the
    /// list removes both.
    static func setWakeHelpers(_ helpers: [WakeHelper], onMachine machineId: String, in root: inout [String: Any]) throws {
        try mutateMachine(machineId, in: &root) { machine in
            guard var wake = machine["wake"] as? [String: Any] else {
                throw EditFailure("\(machineId) has no wake block, so it cannot have wake helpers.")
            }
            if helpers.isEmpty {
                wake.removeValue(forKey: "helpers")
                wake.removeValue(forKey: "helper")
            } else {
                wake["helpers"] = helpers.map { ["machine": $0.machine, "action": $0.action] }
                wake["helper"] = ["machine": helpers[0].machine, "action": helpers[0].action]
            }
            machine["wake"] = wake
        }
    }

    /// Add or replace one site.
    static func upsertSite(_ site: [String: Any], in root: inout [String: Any]) throws {
        guard let id = site["id"] as? String, !id.isEmpty else {
            throw EditFailure("A site needs an id.")
        }
        var sites = (root["sites"] as? [[String: Any]]) ?? []
        if let index = sites.firstIndex(where: { $0["id"] as? String == id }) {
            var merged = sites[index]
            for (key, value) in site {
                if value is NSNull { merged.removeValue(forKey: key) } else { merged[key] = value }
            }
            sites[index] = merged
        } else {
            sites.append(site)
        }
        root["sites"] = sites
    }

    static func removeSite(id: String, in root: inout [String: Any]) throws {
        var sites = (root["sites"] as? [[String: Any]]) ?? []
        let before = sites.count
        sites.removeAll { $0["id"] as? String == id }
        guard sites.count != before else { throw EditFailure("There is no site called \"\(id)\".") }
        // A machine left pointing at a site that no longer exists would fail validation, so the
        // reference goes with it.
        var machines = (root["machines"] as? [[String: Any]]) ?? []
        for index in machines.indices where machines[index]["site"] as? String == id {
            machines[index].removeValue(forKey: "site")
        }
        root["sites"] = sites
        root["machines"] = machines
    }

    /// Name the setup itself.
    static func setSetupName(_ name: String, in root: inout [String: Any]) throws {
        var controller = (root["controller"] as? [String: Any]) ?? [:]
        controller["name"] = name
        root["controller"] = controller
    }

    static func setUpdateRepo(_ repo: String, in root: inout [String: Any]) throws {
        var updates = (root["appUpdates"] as? [String: Any]) ?? [:]
        updates["githubRepo"] = repo
        root["appUpdates"] = updates
    }

    /// Take this device's private settings out of the shared document.
    ///
    /// Never done on its own. `machine.ssh.identityFile` is a path that means something on exactly
    /// one computer and `local` means "whichever device is reading this", so both are wrong in a
    /// document every peer carries — but removing them changes the bytes, and therefore the hash,
    /// and doing that quietly on one device would look like a divergence to everyone else. So it is
    /// an edit a person asks for, and it produces an ordinary new revision.
    static func stripPrivateKeys(in root: inout [String: Any]) throws {
        var machines = (root["machines"] as? [[String: Any]]) ?? []
        for index in machines.indices {
            guard var ssh = machines[index]["ssh"] as? [String: Any] else { continue }
            ssh.removeValue(forKey: "identityFile")
            if ssh.isEmpty {
                machines[index].removeValue(forKey: "ssh")
            } else {
                machines[index]["ssh"] = ssh
            }
        }
        root["machines"] = machines
        root.removeValue(forKey: "local")
    }

    // MARK: - Small helpers

    private static func mutateMachine(
        _ id: String,
        in root: inout [String: Any],
        _ change: (inout [String: Any]) throws -> Void
    ) throws {
        var machines = (root["machines"] as? [[String: Any]]) ?? []
        guard let index = machines.firstIndex(where: { $0["id"] as? String == id }) else {
            throw EditFailure("There is no machine called \"\(id)\" in the file.")
        }
        var machine = machines[index]
        try change(&machine)
        machines[index] = machine
        root["machines"] = machines
    }

    private static func upsertKeyed(
        _ entry: [String: Any],
        in machine: inout [String: Any],
        list: String,
        what: String
    ) throws {
        guard let id = entry["id"] as? String, !id.isEmpty else {
            throw EditFailure("A \(what) needs an id.")
        }
        guard AgentToken.isValid(id) else {
            throw EditFailure("\"\(id)\" is not a usable \(what) id. Use letters, digits and . _ - : @ / \\ ~ = +.")
        }
        var entries = (machine[list] as? [[String: Any]]) ?? []
        if let index = entries.firstIndex(where: { $0["id"] as? String == id }) {
            var merged = entries[index]
            for (key, value) in entry {
                if value is NSNull { merged.removeValue(forKey: key) } else { merged[key] = value }
            }
            entries[index] = merged
        } else {
            entries.append(entry)
        }
        machine[list] = entries
    }
}
